sklearn/utils/estimator_checks.py

```diff
@@ -1433,6 +1435,23 @@ def check_estimators_overwrite_params(name, Estimator):
                      % (name, param_name, original_value, new_value))
 
 
+def check_no_fit_attributes_set_in_init(name, Estimator):
+    """Check that Estimator.__init__ doesn't set trailing-_ attributes."""
+    estimator = Estimator()
+    for attr in dir(estimator):
+        if attr.endswith("_") and not attr.startswith("__"):
+            # This check is for properties, they can be listed in dir
+            # while at the same time have hasattr return False as long
+            # as the property getter raises an AttributeError
+            assert_false(
+                hasattr(estimator, attr),
+                "By convention, attributes ending with '_' are "
+                'estimated from data in scikit-learn. Consequently they '
+                'should not be initialized in the constructor of an '
+                'estimator but in the fit method. Attribute {0!r} '
+                'was found in estimator {1}'.format(estimator, attr))
```

Comment:
Fixed in https://github.com/scikit-learn/scikit-learn/commit/5007e02c88106969553ff453e7e5b66ffa98489f. I added a test while I was at it.